doc/release/upcoming_changes/23240.compatibility.rst

```diff
@@ -0,0 +1,6 @@
+Array-likes that define ``__array_ufunc__`` can now override ufuncs if used as the ``where`` keyword argument
+---------------------------------------------------------------------------------------------------
+If the ``where`` keyword argument of a :class:`numpy.ufunc` is a subclass of
+:class:`numpy.ndarray` or is a duck type that defines
+:func:`numpy.class.__array_ufunc__` it can override the behavior of the ufunc
+using the same mechanism as the input and output arguments.
```

Comment:
So one thing here is that objects like dask or masked arrays may need to unwrap `where=` if it exists.  I suspect `dask` already does (or doesn't even do that anymore).

(I.e. objects doing recursive calls will have to make sure that they also clean up `where=` if it exists.  I suspect all others shouldn't care much, although end-users could.  I doubt that is a problem, but wonder if we should mention it here).

In theory, an object (say dask) could even decide that if it was only passed as `where=` it should just return a NumPy array instead.  In practice, I doubt anyone would bother doing such complicated logic.

In any case, I am happy if Marten is, but we should give e.g. @shoyer (and maybe other stake-holders) a chance to comment.  Maybe even just ping the list briefly.

Reply:
Good point that we should work in the compatibility note that `where` now has to be unwrapped. I don't do this in astropy and may try to quicky run a test with it to see if we would get hit by this.

Reply:
@seberg, good catch on how this will behave a little differently for downstream users. I've added a sentence to the compatibility notes to try and capture what you've explained here.

Reply:
@mhvk, I just ran the astropy tests against this PR and I did get one failure in `astropy.utils.masked.tests.test_function_helpers.TestUfuncLike.test_fix`, not quite sure how to resolve it yet.
```python3
======================================================= FAILURES =======================================================
________________________________________________ TestUfuncLike.test_fix ________________________________________________

self = <astropy.utils.masked.tests.test_function_helpers.TestUfuncLike object at 0x7f260513df90>

    def test_fix(self):
>       self.check(np.fix)

astropy/utils/masked/tests/test_function_helpers.py:644:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
astropy/utils/masked/tests/test_function_helpers.py:68: in check
    o = func(self.ma, *args, **kwargs)
astropy/utils/masked/core.py:841: in __array_function__
    return super().__array_function__(function, types, args, kwargs)
../.local/lib/python3.10/site-packages/numpy/lib/ufunclike.py:53: in func
    return f(x, out=out, **kwargs)
../.local/lib/python3.10/site-packages/numpy/lib/ufunclike.py:118: in fix
    res = nx.floor(x, out=res, where=nx.greater_equal(x, 0))
astropy/utils/masked/core.py:827: in __array_ufunc__
    result = getattr(ufunc, method)(*unmasked, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = MaskedNDArray([[  ———,  True,  True],
               [ True,   ———,  True]]), ufunc = <ufunc 'floor'>
method = '__call__', inputs = (array([[0., 1., 2.],
       [3., 4., 5.]]),)
kwargs = {'where': MaskedNDArray([[  ———,  True,  True],
               [ True,   ———,  True]])}
out = (array([[0., 1., 2.],
       [3., 4., 5.]]),), out_unmasked = (array([[0., 1., 2.],
       [3., 4., 5.]]),)
out_mask = None, out_masks = (None,), d = array([[0., 1., 2.],
       [3., 4., 5.]]), m = None

    def __array_ufunc__(self, ufunc, method, *inputs, **kwargs):
        out = kwargs.pop("out", None)
        out_unmasked = None
        out_mask = None
        if out is not None:
            out_unmasked, out_masks = self._get_data_and_masks(*out)
            for d, m in zip(out_unmasked, out_masks):
                if m is None:
                    # TODO: allow writing to unmasked output if nothing is masked?
                    if d is not None:
>                       raise TypeError("cannot write to unmasked output")
E                       TypeError: cannot write to unmasked output

astropy/utils/masked/core.py:700: TypeError
```

Reply:
`np.fix` was one of the numpy functions that worked directly on the `Masked` class -- but clearly its implementation used `where`. That will be fairly trivial to fix.

So, I don't think that is a show-stopper, but does show that @seberg is correct that not unwrapping `where` arguments is likely the main issue downstream may have. That said, it should mostly be a problem for subclassing `ndarray`.

Reply:
Ya, this is a good demonstration of what can happen in other downstream packages. I'm happy to open a PR in astropy to fix this when the time comes, but hopefully others will chime in about how they think this will affect other packages.

Reply:
Thanks in advance! We're doing overrides now for `Quantity`, `Masked`, and `Distribution`, but the code is fairly similar.

p.s. Inspired by your original use-case, I got at least a little further in updating our `Distribution`... see https://github.com/astropy/astropy/pull/14421 and https://github.com/astropy/astropy/pull/14423

Reply:
Ah, while I may have been right that in some cases unwrapping is necessary. I had a brief look at `dask` and I think it should be fine.  Of course no existing implementation will defer correctly, because unlike `__array_function__` we do not pass all types/objects explicitly.

That means that also here `ndarray.__array_function__` is very much missing a way to defer to the object passed in by `where=`.

The tests work here, but I guess that is just due to the "subclasses before superclasses" rule (did not double check).

So to actually take full effect, all downstream projects need to start checking `where=`  (this makes me wish we had that explicit list like `__array_function__` does...).

Reply:
> That means that also here `ndarray.__array_function__` is very much missing a way to defer to the object passed in by `where=`.
Sorry, I did not follow your logic here, can you explain it a little more? I'm confused since `DuckArray` in `TestImplementingArgs.test_where_duck_array` is not a subclass of `numpy.ndarray`, so I don't see why the "subclasses before superclasses" rule applies.

Reply:
Ah, `__array_ufunc__` works a bit different, since it rarely does a recursive call I think, and using `super().__array_ufunc__` is dubious.
In any case, I think we should change it, although I am not sure how to best test it.

An `__array_ufunc__` implementation normally does something like:
```
def __array_ufunc__(self, args, **kwargs):
    objs = args + kwargs.get(out, ())
    for obj in objs:
        if not is_of_known_type(obj):
             reutn NotImplemented
```
Which arguably should now be:
```
def __array_ufunc__(self, args, **kwargs):
    objs = args + kwargs.get(out, ())
    if "where" in kwargs:
        objs.append(kwargs.get("where"))
```
(or something more pretty, I don't love it, but probably its good)

In any case, we have an `__array_function__` implementation in NumPy and maybe an example in the NEP and both should probably use this pattern?

Reply:
Let's be sure not to confuse `__array_ufunc__` and `__array_function__` - for the last one, there was a systematic go-through of all functions to see what arguments could be arrays.

But @seberg does have a good point: what I think is missing is for `ndarray.__array_ufunc__` to check whether `where` is present, and, like for the inputs and outputs, return `NotImplemented` as needed. (Since astropy typically `super()`s to `ndarray`, that might actually expose a few more problems, but we'll see...). `ndarray.__array_ufunc__` is implemented in `src/multiarray/methods.c`.

Reply:
p.s. Sorry for the duplication, @seberg! But at least I hopefully saved @byrdie from having to find where `ndarary.__array_ufunc__` is defined.  And good point about adjusting examples!

Reply:
Huh, I guess I'm confused why the tests seem to be working now. Do you know of a test I can write that would fail the way the code is now?

Reply:
You may have to use `super().__array_function__` in a subclass.  (And that may always be a bit dubious anyway, see gh-23014... So, in a sense, it maybe doesn't really matter much.

Reply:
I think we need a test for an ndarray subclass that overrides `__array_ufunc__`  and by mistake passes its own instance down as where, like,
```
class Mine(np.ndarray):
    def __array_ufunc__(...):
        ...
        return super().__array_ufunc__(..., where=<instance-of-Mine>)
```

Maybe the best place is (near) `test_ufunc_override_with_super` in `test_umath.py`?

Reply:
@mhvk, I'm trying to implement a test like you suggested above, but it's leading to infinite recursion since `Mine.__array_ufunc__` is basically a no-op, and then `ndarray.__array_ufunc__` sees the instance of `Mine` and calls `Mine.__array_ufunc__` again. Is this what you expected to happen?

Reply:
You are getting the recursion before any change in `ndarray.__array_ufunc__`, right? If so, that's not what I initially expected, but good in the sense that it means nobody can have made the mistake of passing in their subclass to `where` and not notice!

But with a change in `ndarray.__array_ufunc__`, that should return `NotImplemented` and then the result should be a `TypeError` (which is what the test would check).

Reply:
Yeah, no change in `ndarray.__array_ufunc__`. Sounds good though, thanks for walking me through this!

Reply:
@seberg and @mhvk, I think I've sorted this out. I've modified `ndarray.__array_ufunc__` to return `NotImplemented` if `where` overrides `__array_ufunc__`, added a test along the lines suggested by @mhvk, and updated the NEP to reflect this pattern. Let me know what you think.